### ISSUE TITLE
availability of trade profit while generating sell signals

### ISSUE THREAD

**Issue by aleeepp** [freqtrade/freqtrade#4276]
Hi first. I am sorry for my english i know it's not good. I am 32 years old and my hobby is trading. Its a hobby because i won a lot more with my company on accountability. 

For the last 2 years  I was away from trading because I was studying an MBA. But now i finished and for the last 3 moth i was studying  freqtrade. My last bot was wiht Gekko.

Finished that little intro. I want to say Freqtrade has 2 problems for me:

1) I have 3 signal for buy and I want to use different strategy  for sell For example if i buy with the signal A i will use the logic C for sell

2) This is the worst problem: I need to control the sell signal with the profit.
For example if my profit of the current trade is >3% I will use the logic A, but if my profit is between 0 and 3% I will use a different strategy.
And if my profit an hour ago was negative i will use another strategy

1/2 its necessary to use on backtest (if you can't test the strategy you can't go to productive)

This is my see point . Freqtrade  is a great work !! But need to have more flexibility

**Comment by xmatthias** [comment]
First of all, please pick a better title next time.
"freqtrade's worst problem" is not a valid title, and i'll simply close the next issue if you can't come up with something clearly identifying it's content.
it may be the worst problem for you, but it's really not a problem, but an enhancement request (and one we will probably not be able to fullfill).

---

what you're asking for is probably not possible without sacrificing the backtesting speed (which is however not an option, as it'll force us to remove hyperopt).

Hyperopt is ONLY made possible by having a quick (vectorized) backtesting.
This means indicators (and timeframes) are calculated across the whole timerange at once. obviously, you'll need to be careful with "lookahead" bias, but it's worth the tradeoff to backtest a year of 1m data in under 1 minute.

If you need to know the trade's profit while generating the sell signal, we'd be enforced to drop this method, and instead use a loop to generate the signals.
This would make backtesting VERY slow (basically unusable, at least for hyperopting) - and would therefore force us to remove hyperopt (which relies on running backtesting quickly).
As hyperopt is however a core feature, i don't think that's a route we'll go anytime soon, especially not to facilitate a request that's most likely not going to be used by 90% of the people, as it's to facilitate one random strategy (out of many).


for live / dry run, you can do this already by combining some of the methods [documented here](https://www.freqtrade.io/en/latest/strategy-customization/#additional-data-dataprovider) - but as you correctly say, that's not backtestable so you will need to rely on dry-runs to confirm your logic.


You might try to go with [custom stoploss](https://www.freqtrade.io/en/latest/strategy-advanced/#custom-stoploss) - which while not acting as sell signal, will still allow you to pull stoploss rather close, so it'll sell the next time profit drops below this.

**Comment by aleeepp** [comment]
Hi xmatthias, thanks for your response! First sorry for my title after my MBA i am bit more "salesman" :) 
Why you say only the 10% will use this feature? I don't understand who can put a signal sell without knowing your current profit. 

But I understand your point and I know Freqtrade  is a great product!

Is it possible to know on custom stop loss the trend or change on price? Why -> because for example I will sell my position if after 3h my profit is between 1 and 3% but ONLY if I am losing profit. If a 30 min ago I was -1% and now (3h) is between 1 / 3% I won't sell because the trend is up, and I can have the position a bit more.

**Comment by jeroenpost86** [comment]
I was wondering something similair.

An option in the backtest functionality to have it iterate over the data and feed it to the strategy like it would in dry-run or live mode would be very useful.

This is a basic strategy I came up with to do part of what aleeep was asking for: Buy, sell when there is profit, wait until price goes down to sell again, etc.
Because there are many variables and calls to the database this would be very slow to test, but having that option would be great. 

My suggestion is having a parameter in the backtest function to enable simple iterations. I would not mind having to wait an hour to test a month worth of data, it would still be very useful to be able to test spaghetti code like this, for people like me who don't know python or want to do crazy things without messing with the rest of the source code.


```

from freqtrade.strategy.interface import IStrategy
from typing import Dict, List
from functools import reduce
from pandas import DataFrame
from freqtrade.persistence import Trade
from datetime import datetime, date, timedelta


import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib
import numpy  # noqa
import logging

logger = logging.getLogger(__name__)

class jeroen_test(IStrategy):

    # Minimal ROI designed for the strategy.
    minimal_roi = {
        "0": 0.02
    }

    order_types = {
        'buy': 'market',
        'sell': 'market',
        'stoploss': 'market',
        'stoploss_on_exchange': False
    }

    # Optimal stoploss designed for the strategy
    stoploss = -10

    # Optimal timeframe for the strategy
    timeframe = '1m'

    def calc_profit(self, price: float, current: float) -> float:
        fee = 1.0007
        profit = ((current*fee) -
            (price*fee))

        return float(f"{profit:.8f}")

    def calc_percentage_lower(self, price: float, current: float) -> float:
        fee = 1.0007
        price = price*fee
        current = current*fee
        lowerpercent = ((price-current)/(price*fee))*100

        return float(f"{lowerpercent:.8f}")

    def bot_loop_start(self, **kwargs) -> None:
        print(" ")

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """ Adds several different TA indicators to the given DataFrame
        """
        profit = False
        profit_percent = False
        percent_lower = False
        current_price = dataframe['close'].iloc[-1]

        dataframe['should_sell'] = False
        dataframe['should_buy'] = False

        # Get the previous trade
        trade = Trade.get_trades_proxy(is_open=False, pair=metadata['pair'])
        if trade:
            trade = trade[-1]
            lsp = trade.close_rate
            if lsp:
                percent_lower = self.calc_percentage_lower(price=lsp, current=current_price)
            # Found a bug? When force selling it doesnt close it
            else:
                lsp = trade.open_rate
                if lsp:
                    percent_lower = self.calc_percentage_lower(price=lsp, current=current_price)
        else:
            lsp = 0.00

        # Get the current Trade
        trade = Trade.get_trades_proxy(is_open=True, pair=metadata['pair'])
        if trade:
            trade = trade[-1]
            lbp = trade.open_rate
            open_trade = True
            profit = self.calc_profit(price=lbp, current=current_price)
            profit_percent = (profit/lbp)*100
        else:
            lbp = 0.00
            open_trade = False
            profit = False
            profit_percent = False


        print("------------")

        print("Last Sold For:", lsp)

        if open_trade:
            print("Bought for: ", lbp)
        print("Current Price: ", current_price)
        if profit:
            print("Current Profit: ", profit, " ", float(f"{profit_percent:.8f}"), "%")
        if percent_lower and not open_trade:
            print("Percent Lower: ", float(f"{percent_lower:.8f}"), "%")


        # Should we Sell?
        if profit_percent:
            if profit_percent > 1:
                dataframe['should_sell'] = True

        # Should we buy?
        if not open_trade:
            if (lsp == 0.00 ) & (lbp == 0.00):
                dataframe['should_buy'] = True
            # Is the percentage of what we sold for and the current price 2% lower
            if percent_lower > 2:
                dataframe['should_buy'] = True


        dataframe['last_sell_price'] = lsp
        dataframe['last_buy_price'] = lbp


        print("Current Dataframe:")
        print(dataframe.tail(1))


        return dataframe

    def populate_buy_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[(
            # We have not bought or sold anything yet, lets buy!
            ((dataframe['last_sell_price'] == 0.00) & (dataframe['last_buy_price'] == 0.00) ) |
            (
                # Make sure the last selling price is higher than the current price
                ((dataframe['last_sell_price']) > dataframe['close']) &

                # Calculated earlier
                (dataframe['should_buy'] == True)
            )
            ), 'buy'
        ] = 1

        return dataframe

    def populate_sell_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[(
                # Make at least profit
                (dataframe['last_buy_price'] < (dataframe['close'])) &

                # Calculated earlier
                (dataframe['should_sell'] == True) &

                # If we have nothing we bought, there is nothing to sell
                (dataframe['last_buy_price'] > 0.00)
            ), 'sell'
        ] = 1

        return dataframe
```
